### PR TITLE
Implement MiR Fleet Adapter

### PR DESCRIPTION
**Pull request by methylDragon**
Implements MiR fleet adapter using the RMF Fleet Adapter Python bindings

### CODE REVIEW

**Review by mxgrey — CHANGES_REQUESTED** [review]
I think this implementation is conceptually really great, and I'm very grateful for the amount of thought that went into it.

I do have some concerns about the complexity of the state management. Of course I understand that this complexity was necessary in order to interface correctly with the fleet adapter API, but my concern is that we don't have a very good way of testing this complexity without interfacing with a real MiR server.

I think in the long run, we'll want to refactor migrate this work into the ongoing free fleet effort. That way we can isolate the state management logic from the MiR API and do proper unit testing. Then I'll feel much more comfortable with the level of complexity here.

I've left some comments for things that look like they could be errors. Some of those comments might be incorrect, since I might just be misunderstanding some details of the implementation.

File: fleet_adapter_mir/fleet_adapter_mir.py [hunk 1]
```diff
@@ -0,0 +1,827 @@
...
+        # Construct remaining path list
+        for waypoint in waypoints[::-1]:
+            # Split timestamp into decimal and whole second portions
+            _sub_seconds, _seconds = math.modf(waypoint.time.timestamp())
+
+            _msg = Location()
+            _msg.x, _msg.y, _msg.yaw = waypoint.position
+            _msg.t.sec, _msg.t.nanosec = int(_seconds), int(_sub_seconds * 1e9)
+
+            self.rmf_remaining_path.append(_msg)
```
[the remark this answers is not in the document]
Comment (methylDragon) [reply fleet_adapter_mir/fleet_adapter_mir.py]: Whoops! I mixed up another class variable. I'm supposed to be populating an array of Locations to feed into the RobotState message for publishing the FleetState in a downstream script (e.g. in `main.py`)

Also, a note on the robot state message path member. In ROS1 I noticed that paths are in [last, ..., first] order. Is that the same for RMF? (Notably, the RMF plan waypoints are given in [first, ..., last] order)

Updated: 5f6bd6e37cd3c7f403671bd4c4e7ed643932da1b

File: fleet_adapter_mir/fleet_adapter_mir.py [hunk 2]
```diff
@@ -0,0 +1,827 @@
...
+
+        # SEMANTIC STATE INFERENCE AND ADJUSTMENT =============================
+        # See docs for more information on the state transitions
+
+        # MiR_Move: Non-RMF Commanded Move-To-Coordinate
+        # (0, 1, 0) and (1, 0 ,0) --> (0, 0, 0)
+        # When robot is done moving, robot will be IdleAtUnknown
+        if not self.rmf_path_requested and robot_mode == RobotMode.MODE_MOVING:
+            # Unset all
+            self.rmf_current_lane_index = 0
```
[the remark this answers is not in the document]
Comment (methylDragon) [reply fleet_adapter_mir/fleet_adapter_mir.py:707]: Updated: 44660a25edde4127f610b6d64508b2836805fd4e

File: fleet_adapter_mir/fleet_adapter_mir.py [hunk 3]
```diff
@@ -0,0 +1,827 @@
...
+            self.rmf_updater.update_position(self.rmf_current_waypoint_index,
+                                             self.rmf_ori)
+        # In Transit or Idle in Lane
+        # States: (1, 0, 0), (1, 0, 1)
+        elif self.rmf_current_lane_index:
+            self.rmf_updater.update_position(rmf_3d_pos,
+                                             self.rmf_current_lane_index)
+        # In Transit, Unknown Lane
+        # States: (0, 0, 1)
+        elif self.rmf_target_waypoint_index:  # In Unknown Lane
```
[the remark this answers is not in the document]
Comment (methylDragon) [reply fleet_adapter_mir/fleet_adapter_mir.py]: Updated: dad547bb9bc3b84b366f3b70e4b81e04243704b9

File: fleet_adapter_mir/fleet_adapter_mir.py [hunk 4]
```diff
@@ -0,0 +1,827 @@
...
+        rmf_3d_pos = [*rmf_pos, rmf_ori]
+
+        # At waypoint
+        # States: (0, 1, 0)
+        if self.rmf_current_waypoint_index:
+            self.rmf_updater.update_position(self.rmf_current_waypoint_index,
+                                             self.rmf_ori)
+        # In Transit or Idle in Lane
+        # States: (1, 0, 0), (1, 0, 1)
+        elif self.rmf_current_lane_index:
```
[the remark this answers is not in the document]
Comment (methylDragon) [reply fleet_adapter_mir/fleet_adapter_mir.py]: Updated: dad547bb9bc3b84b366f3b70e4b81e04243704b9

File: fleet_adapter_mir/fleet_adapter_mir.py [hunk 5]
```diff
@@ -0,0 +1,827 @@
...
+
+        rmf_pos = self.transforms['mir_to_rmf'].transform(mir_pos)
+        rmf_ori = (math.radians(mir_ori % 360)
+                   + self.transforms['mir_to_rmf'].get_rotation())
+
+        rmf_3d_pos = [*rmf_pos, rmf_ori]
+
+        # At waypoint
+        # States: (0, 1, 0)
+        if self.rmf_current_waypoint_index:
```
[the remark this answers is not in the document]
Comment (methylDragon) [reply fleet_adapter_mir/fleet_adapter_mir.py]: Updated: dad547bb9bc3b84b366f3b70e4b81e04243704b9

File: main.py [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+                continue
+            else:
+                if recursive:
+                    val = sanitise_dict(val, False, True)
+                output[key] = val
+
+    return output
+
+
+def create_nav_graph(config):
```
[the remark this answers is not in the document]
Comment (methylDragon) [reply main.py]: Good point, updated: cd2cd394cbbace544e5cfbe073c1aa002a2f9967

File: fleet_adapter_mir/fleet_adapter_mir.py [hunk 7]
```diff
@@ -0,0 +1,827 @@
...
+                    next_mission_wait = (ros_waypoint_leave_time - ros_now)
+                else:
+                    # Prevent spinning out of control when paused
+                    self._path_quit_cv.acquire()
+                    self._path_quit_cv.wait(1)
+                    self._path_quit_cv.release()
+
+                # CHECK FOR PRE-EMPT ==========================================
+                if self._path_quit_event.is_set():
+                    self.stop()
```
[the remark this answers is not in the document]
Comment (methylDragon) [reply fleet_adapter_mir/fleet_adapter_mir.py:273]: I don't understand :O

This section is explicitly for pre-empting/killing the thread (when the quit event flag is set). So new path_closure threads are supposed to be spun up before this call to stop can happen.

Before any new threads are spun up, old threads are killed by a call to [`self.stop()`](https://github.com/osrf/fleet_adapter_mir/pull/1/files/7dcb4d5d46ce188e8fa1f28814ceb50f3798daef#diff-1d43fe5e5105e78b185ef39c7d4868f4R209)

```
    def follow_new_path(self,
                        waypoints,
                        next_arrival_estimator,  # function!
                        path_finished_callback):
        self.stop()  # <--- here
        self.current_task_id += 1
```

And `self.stop()` blocks until threads are killed. (Due to the `join()` call.)

I am hesitant to

[the remark this answers is not in the document]
Comment (methylDragon) [reply fleet_adapter_mir/fleet_adapter_mir.py:273]: Oh, actually I see the issue.

Hopefully this wait fixes it? 8b8dd79467253e3e6e977ee9c653ea72266230bd

**Comment by mxgrey** [comment]
To move this review along, I'm going to merge it into a branch of this repo and open a new PR into main so I can push changes directly.

**Pull request merged**